### PR TITLE
Example/token translation

### PR DESCRIPTION
**Pull request by InteractiveTimmy**
# Description

The scope of the changes in this Pull Request are to add core scripts, along with the ability to **automatically update** tokens based on the latest design articles. This is currently only inclusive of `core/color/decorative` and `core/color/functional` color tokens, but is expandable to include other automation tokens based on what is readily available.

### CODE REVIEW

File: scripts/commands/update/update.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+ */
+const update = () => {
+  const next = {
+    functional: {
+      color: new ColorToken({
+        format: ColorToken.CONSTANTS.TOKEN_FORMATS.AUTOMATED,
+        category: ColorToken.CONSTANTS.CATEGORIES.COLOR,
+        data: nextTokens.core["core color"],
+        name: "functional",
+      }).normalize(),
```
Comment (cscheffauer) [inline scripts/commands/update/update.js:26]: why is the normalize necessary here?

File: scripts/models/color-token/color-token.js [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+  /**
+   * Merge a ColorToken into an existing ColorToken.
+   *
+   * @param {DTO} dto
+   * @param {ColorToken} dto.source - ColorToken to merge into the destination ColorToken.
+   * @param {ColorToken} dto.destination - ColorToken to be receiving a merge.
+   * @returns {ColorToken} - Merged ColorToken.
+   */
+  static merge({ source, destination }) {
+    if (source.category !== Token.CONSTANTS.CATEGORIES.COLOR || destination.category !== Token.CONSTANTS.CATEGORIES.COLOR) {
```
Comment (cscheffauer) [inline scripts/models/color-token/color-token.js:86]: i like the sanity checks here - nice work!

File: scripts/models/color-token/color-token.js [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    );
+  }
+
+  /**
+   * Merge a ColorToken into an existing ColorToken.
+   *
+   * @param {DTO} dto
+   * @param {ColorToken} dto.source - ColorToken to merge into the destination ColorToken.
+   * @param {ColorToken} dto.destination - ColorToken to be receiving a merge.
+   * @returns {ColorToken} - Merged ColorToken.
```
Comment (cscheffauer) [inline scripts/models/color-token/color-token.js:83]: probably worth adding here that the returned ColorToken is normalised

File: scripts/models/color-token/color-token.js [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+    switch (from) {
+      case Token.CONSTANTS.TOKEN_FORMATS.STANDARD:
+        final = grades;
+        break;
+
+      case Token.CONSTANTS.TOKEN_FORMATS.AUTOMATED:
+        final = Object.entries(grades).reduce((formatted, [key, value]) => {
+          const mutable = { ...formatted };
+          const grade = key.split("-").pop();
```
Comment (cscheffauer) [inline scripts/models/color-token/color-token.js:128]: so the last string in the key is the grade? would be nice to add a short comment here why that is the case

File: scripts/models/color-token/color-token.js [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+          const { r, g, b, a } = value.rgba;
```
Comment (cscheffauer) [inline scripts/models/color-token/color-token.js:129]: what happens if there is no rgba value? do u want to provide a fallback for that here?

File: scripts/models/token/token.constants.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
+/**
+ * Supported token categories.
+ */
+const CATEGORIES = {
+  COLOR: "color",
+};
+
+/**
+ * Specificaions for token serials.
```
Comment (cscheffauer) [inline scripts/models/token/token.constants.js:9]: typo

File: scripts/models/token/token.constants.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+ */
+const SERIAL_SPECIFICATIONS = {
+  INDENT: 2,
```
Comment (cscheffauer) [inline scripts/models/token/token.constants.js:12]: not quite sure what this constant here is for. what is a SERIAL_SPECIFICATIONS.INDENT representing? like how can the number 2 here be explained?

File: scripts/utils/sort-object/sort-object.constants.js [hunk 8]
```diff
@@ -0,0 +1,8 @@
+/**
+ * Special character used to prevent JavaScript auto-sorting in non-serialized JSON Objects.
+ */
+const SORT_LOCK_CHARACTER = "�";
```
Comment (cscheffauer) [inline scripts/utils/sort-object/sort-object.constants.js:4]: not quite understanding the usage of this - if that is amended, u stop sorting everything underneath? is that correct?

**Review by cscheffauer — CHANGES_REQUESTED** [review]
haven't tested locally yet. will need more time to check
looks really good at a first glance (just looking at the code itself, without testing functionality)

**Comment by torgeadelin** [comment]
I would love to write typescript going further!

File: scripts/commands/update/update.js (hunk 1 above)
Comment (InteractiveTimmy) [reply scripts/commands/update/update.js:26]: `new ColorToken({ /* ... */ }).normalize()` converts the token immediately to `standard` format, which follows the formatting of the local token files. In some cases [like with gradients], tokens need additional steps if they aren't standardized/normalized, which is why this isn't amended as a part of the constructor flow.

File: scripts/models/color-token/color-token.js (hunk 3 above)
Comment (InteractiveTimmy) [reply scripts/models/color-token/color-token.js:83]: Will do! Updating the JSDoc now.

File: scripts/models/color-token/color-token.js (hunk 4 above)
Comment (InteractiveTimmy) [reply scripts/models/color-token/color-token.js:128]: Will add a comment!

File: scripts/commands/update/update.js (hunk 1 above)
Comment (cscheffauer) [reply scripts/commands/update/update.js:26]: sounds good to me - i think its worth indicating that somehow (we don't have TS yet, which would have been helpful here, so a comment or something I guess would be good) :)

File: scripts/models/color-token/color-token.js (hunk 5 above)
Comment (InteractiveTimmy) [reply scripts/models/color-token/color-token.js:129]: The exporter, `momentum-abstract`, converts all color token values to `hex24`, `rgba`, `lch`, and `cmyk`. The most reliable, and platform adopted standard, is the `rgba` value. When provided with a color token of format `automation`, this value always exists.

I can add a flow/check if you believe it's still necessary.

File: scripts/models/token/token.constants.js (hunk 7 above)
Comment (InteractiveTimmy) [reply scripts/models/token/token.constants.js:12]: Serial specifications is the "shape" of the serial generated by `new Token({ /* ... */ }).serial`. I'll add a comment around this, and potentially change the name to better reflect the intent.

File: scripts/utils/sort-object/sort-object.constants.js (hunk 8 above)
Comment (InteractiveTimmy) [reply scripts/utils/sort-object/sort-object.constants.js:4]: JavaScript automatically sorts object keys.

If a JavaScript object is sorted via a reducer to be:

```js
const obj = {
  '00': 'value',
  '05': 'value',
  '50': 'value',
  '100': 'value',
};
```

When stringifying or just printing the line to console, it appears as so:

```js
console.log(JSON.stringify(obj, null, 2));

// output
{
  "50": "value",
  "100": "value",
  "00": "value",
  "05": "value"
}
```

For some reason, the keys that start with `0` get pushed to the end. Feel free to test this locally 😉

**Comment by InteractiveTimmy** [comment]
> I would love to write everything in typescript going further!

I strongly agree! Since the project didn't include any TS/TSC at the time of writing this within it's time-box, the call was made to introduce this in native **CommonJS** to reduce implementation time. There is future work that we can execute on in regards to this, which will be migrating this to TS once the automation triggers are ready.

File: scripts/commands/update/update.js (hunk 1 above)
Comment (InteractiveTimmy) [reply scripts/commands/update/update.js:26]: The JSDoc indicates this currently:

```js
  /**
   * Normalizes this ColorToken into the project's Standard format.
   *
   * @returns {this} - This color token.
   */
  normalize() {
    return this.updateFormat({ format: Token.CONSTANTS.TOKEN_FORMATS.STANDARD });
  }
  ```
  
  Let me know if this isn't enough.
  
  >we don't have TS yet, which would have been helpful here, so a comment or something I guess would be good
  
  This is a very valid point, TypeScript would have provided a resolution to this comment.

**Review by cscheffauer — APPROVED** [review]

**Pull request merged**